xdsl/dialects/scf.py

```diff
@@ -148,9 +155,13 @@ def get(
 class For(IRDLOperation):
     name = "scf.for"
 
-    lb: Operand = operand_def(IndexType)
-    ub: Operand = operand_def(IndexType)
-    step: Operand = operand_def(IndexType)
+    T = Annotated[
+        Attribute, AnyOf([IndexType, SignlessIntegerConstraint]), ConstraintVar("T")
+    ]
+
+    lb: Operand = operand_def(T)
+    ub: Operand = operand_def(T)
+    step: Operand = operand_def(T)
```

Comment:
I initially wanted to do

```suggestion
    step: Operand = operand_def(signlessIntegerLike)
```

from [here](https://github.com/xdslproject/xdsl/blob/79bbb8b2ca2b23c03636bb216206b2193fd74635/xdsl/dialects/arith.py#L39), but that allowed signed ints.
So, now I'm also questioning the intention behind that type alias.

Reply:
@math-fehr ?

Reply:
I am not quite sure I am getting the question, but your change would make it such that the operands could have different types, whereas the current version enforces them to all be the same (at least AFAIU).

Reply:
@webmiche my bad, I didn't select the right range of LoCs (updated above); I meant if I change all three to be restricted to `signlessIntegerLike`, signed integers (i.e., `si32`) are allowed.

Reply:
I'm not sure what you mean, but with your current change, you should get what you want.
Which is three operands that have the same type, and that type being either an index type, or a signless integer.

Reply:
Hmm, I am not 100% sure anymore about how PyRDL will handle this, so maybe I am just wrong here :sweat_smile: 
But intuitively, changing all of them to `singlessIntegerLike` might make it possible to instantiate one with `i32` and another one with `i64`, which would not make sense to me. Having the type alias to me looks like all of them should be instantiated to the same type.
I might be completely wrong here, I am not up to speed with these things anymore, so feel free to point it out :)

Also, if this change really makes all the types the same, how would you then encode it if you have multiple operands with potentially different types?